### PR TITLE
isotonic regression with PAV

### PR DESCRIPTION
**Pull request by michaelsdr**
Isotonic regression using PAV algorithm with a pure_callback() to execute PAV on the CPU using numba

### CODE REVIEW

File: jaxopt/_src/isotonic.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+@njit
+def _isotonic_l2_pav(y):
+    """Solves an isotonic regression problem using PAV.
+
+    Formally, it solves argmin_{v_1 >= ... >= v_n} 0.5 ||v - y||^2.
+
+    Args:
+      y: input to isotonic regression, a 1d-array.
+      sol: where to write the solution, an array of the same size as y.
+    """
```
[the remark this answers is not in the document]
Comment (fabianp) [reply jaxopt/_src/isotonic.py:29]: note: I would be also fine to delete this docstring since it's a private method

File: tests/isotonic_test.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import numpy as onp
+
+from jax.test_util import check_grads
+from jaxopt._src.isotonic import isotonic_l2_pav
+from jaxopt._src import test_util
+
+
+class IsotonicPavTest(test_util.JaxoptTestCase):
+
+    def test_output_shape_and_dtype(self):
```
Comment (fabianp) [inline tests/isotonic_test.py:29]: add docstring to tests (could be a single phrase): https://g3doc.corp.google.com/company/readability/home/acquiring-readability/python-readability-advice.md?cl=head#test_docstrings

File: tests/isotonic_test.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        n = 10
+        y = jax.random.normal(jax.random.PRNGKey(0), (n,))
+        output =  isotonic_l2_pav(y)
+        assert output.shape == y.shape
+        assert output.dtype == y.dtype
+
+    def test_gradient(self):
```
Comment (fabianp) [inline tests/isotonic_test.py:36]:
```suggestion
    def test_gradient(self, n = 10):
```
and remote the line below. This way its clear that the test could be run with other values of n

Comment (fabianp) [reply tests/isotonic_test.py:36]: same for the test below

File: tests/isotonic_test.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+from jaxopt._src import test_util
+
+
+class IsotonicPavTest(test_util.JaxoptTestCase):
+
+    def test_output_shape_and_dtype(self):
+        n = 10
+        y = jax.random.normal(jax.random.PRNGKey(0), (n,))
+        output =  isotonic_l2_pav(y)
+        assert output.shape == y.shape
```
Comment (mblondel) [inline tests/isotonic_test.py:33]: Use `self.assertEqual(., .)`

File: tests/isotonic_test.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            return (isotonic_l2_pav(y**3) + isotonic_l2_pav(y)**2).mean()
+
+        check_grads(l, (y, ), order=2)
+
+    def test_vmap(self):
+        b = 16
+        n = 10
+        y = jax.random.normal(jax.random.PRNGKey(0), (b, n))
+        isotonic_l2_pav_vmap = jax.vmap(isotonic_l2_pav)
+        assert jnp.max(jnp.absolute(isotonic_l2_pav_vmap(y)[0] - isotonic_l2_pav(y[0]))) == 0
```
Comment (mblondel) [inline tests/isotonic_test.py:50]: Same.

File: tests/isotonic_test.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
```
Comment (mblondel) [inline tests/isotonic_test.py:51]: Remove blanks lines.

**Review by mblondel — APPROVED** [review]

File: jaxopt/_src/isotonic.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: jaxopt/_src/isotonic.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+import numpy as np
+import jax
+import jax.numpy as jnp
+
+# pylint: disable=g-import-not-at-top
+try:
+    from numba import njit
+except ImportError:
+    warnings.warn("Numba could not be imported. Code will run much more slowly."
+                  " To install, please run 'pip install numba'.")
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: jaxopt/_src/isotonic.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+@jax.custom_jvp
+def isotonic_l2_pav(y):
+    """Solves an isotonic regression problem using PAV.
+
+    Formally, it solves argmin_{v_1 >= ... >= v_n} 0.5 ||v - y||^2.
+
+    Args:
+      y: input to isotonic regression, a 1d-array.
+      sol: where to write the solution, an array of the same size as y.
+    """
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py:101]: Done

File: tests/isotonic_test.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+
+
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply tests/isotonic_test.py]: Done

File: tests/isotonic_test.py (hunk 2 above)
Comment (michaelsdr) [reply tests/isotonic_test.py]: Done

File: jaxopt/_src/isotonic.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
+"""Isotonic optimization routines in Numba with a Jax wrapper and custom JVP."""
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: jaxopt/_src/isotonic.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: jaxopt/_src/isotonic.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+import warnings
+import numpy as np
+import jax
+import jax.numpy as jnp
+
+# pylint: disable=g-import-not-at-top
+try:
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py:24]: Done

File: jaxopt/_src/isotonic.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    warnings.warn("Numba could not be imported. Code will run much more slowly."
+                  " To install, please run 'pip install numba'.")
+
+    # If Numba is not available, we define a dummy 'njit' function.
+    def njit(func):
+        return func
+
+
+@njit
+def _isotonic_l2_pav(y):
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py:36]: Done

File: jaxopt/_src/isotonic.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    """Solves an isotonic regression problem using PAV.
+
+    Formally, it solves argmin_{v_1 >= ... >= v_n} 0.5 ||v - y||^2.
+
+    Args:
+      y: input to isotonic regression, a 1d-array.
+      sol: where to write the solution, an array of the same size as y.
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: jaxopt/_src/isotonic.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
+"""Isotonic optimization routines in Numba with a Jax wrapper and custom JVP."""
+
+import warnings
+import numpy as np
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: jaxopt/_src/isotonic.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+@jax.custom_jvp
+def isotonic_l2_pav(y):
+    """Solves an isotonic regression problem using PAV.
+
+    Formally, it solves argmin_{v_1 >= ... >= v_n} 0.5 ||v - y||^2.
+
+    Args:
+      y: input to isotonic regression, a 1d-array.
+      sol: where to write the solution, an array of the same size as y.
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: jaxopt/_src/isotonic.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+    i = 0
+    while i < n:
+        k = target[i] + 1
+        sol[i + 1: k] = sol[i]
+        i = k
+    return sol.astype(y.dtype)
+
+
+@jax.custom_jvp
+def isotonic_l2_pav(y):
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py:91]: Done

File: jaxopt/_src/isotonic.py [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    """
+    # Define the expected shape & dtype of output.
+    result_shape_dtype = jax.ShapeDtypeStruct(
+        shape=y.shape,
+        dtype=y.dtype)
+    return jax.pure_callback(_isotonic_l2_pav, result_shape_dtype, y)
+
+
+def _jvp_isotonic_l2_jax_pav(solution, vector, eps=1e-8):
+    """Jacobian at solution of the isotonic regression times vector product."""
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: jaxopt/_src/isotonic.py [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+    A = jnp.cumsum(one_hot_start, axis=-1)
+    A = jnp.append(jnp.diff(A[::-1], axis=0)[::-1],
+                   A[-1].reshape(1, -1), axis=0)
+    B = A.copy()
+    return (((B.T * (B @ vector)).T) / (A.sum(1, keepdims=True) + 1e-8)).sum(0)
+
+
+@isotonic_l2_pav.defjvp
+def isotonic_l2_pav_jvp(primals, tangents):
+    """Jacobian at solution of the isotonic regression solved with PAV times vector product.
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: tests/isotonic_test.py (hunk 4 above)
Comment (michaelsdr) [reply tests/isotonic_test.py]: Done

File: tests/isotonic_test.py (hunk 5 above)
Comment (michaelsdr) [reply tests/isotonic_test.py]: Done

File: tests/isotonic_test.py (hunk 6 above)
Comment (michaelsdr) [reply tests/isotonic_test.py]: Done

File: jaxopt/_src/isotonic.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+    """Solves an isotonic regression problem using PAV.
+
+    Formally, it solves argmin_{v_1 >= ... >= v_n} 0.5 ||v - y||^2.
+
+    Args:
+      y: input to isotonic regression, a 1d-array.
+      sol: where to write the solution, an array of the same size as y.
+    """
+    # Define the expected shape & dtype of output.
+    result_shape_dtype = jax.ShapeDtypeStruct(
```
[the remark this answers is not in the document]
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: jaxopt/_src/isotonic.py [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Isotonic Regression.
+
+Original code from:
+Fast Differentiable Sorting and Ranking
+Mathieu Blondel, Olivier Teboul, Quentin Berthet, Josip Djolonga
+In proceedings of ICML 2020 arXiv:2002.08871
```
Comment (fabianp) [inline jaxopt/_src/isotonic.py:21]: move this to somewhere inside the actual functions (cite it as "see section Y in Blondel et al.")

Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: tests/isotonic_test.py [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+        y = jax.random.normal(jax.random.PRNGKey(0), (n,))
+        output = isotonic_l2_pav(y)
+        self.assertEqual(output.shape, y.shape)
+        self.assertEqual(output.dtype, y.dtype)
+
+    def test_gradient(self, n=10):
+        """Checks the gradient with finite differences."""
+        y = jax.random.normal(jax.random.PRNGKey(0), (n,))
+
+        def l(y):
```
Comment (fabianp) [inline tests/isotonic_test.py:42]: please use more specific names, such as `loss`

File: tests/isotonic_test.py [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+            return (isotonic_l2_pav(y**3) + isotonic_l2_pav(y)**2).mean()
+
+        check_grads(l, (y, ), order=2)
+
+    def test_vmap(self, n=10, b=16):
```
Comment (fabianp) [inline tests/isotonic_test.py:47]: suggestion: use a more descriptive name for n and b (by reading the code it's not clear what they refer to)

File: tests/isotonic_test.py [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+        """Verifies vmap."""
+        y = jax.random.normal(jax.random.PRNGKey(0), (b, n))
+        isotonic_l2_pav_vmap = jax.vmap(isotonic_l2_pav)
+        for i in range(b):
+            self.assertEqual(jnp.max(jnp.absolute(
```
Comment (fabianp) [inline tests/isotonic_test.py:52]: suggestion: use multiple lines to make this expression more readable

Comment (michaelsdr) [reply tests/isotonic_test.py:52]: Done

File: tests/isotonic_test.py (hunk 22 above)
Comment (michaelsdr) [reply tests/isotonic_test.py:47]: Done thanks

File: tests/isotonic_test.py (hunk 21 above)
Comment (michaelsdr) [reply tests/isotonic_test.py:42]: Done

File: jaxopt/_src/isotonic.py [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+
+  Formally, it solves argmin_{v_1 >= ... >= v_n} 0.5 ||v - y||^2.
+
+  Args:
+    y: input to isotonic regression, a 1d-array.
+
+  Returns:
+    The solution, an array of the same size as y.
+  """
+  # Define the expected shape & dtype of output.
```
Comment (mblondel) [inline jaxopt/_src/isotonic.py:103]: Move this comment below the if

File: tests/isotonic_test.py [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    output = isotonic_l2_pav(y)
+    self.assertEqual(output.shape, y.shape)
+    self.assertEqual(output.dtype, y.dtype)
+
+  def test_compare_with_sklearn(self, n=10):
+    """Compares the output with the one of sklearn."""
+    y = jax.random.normal(jax.random.PRNGKey(0), (n,))
+    output = isotonic_l2_pav(y)
+    output_sklearn = jnp.array(isotonic.isotonic_regression(y, increasing=False))
+    self.assertAlmostEqual(((output - output_sklearn)**2).mean().max(), 0)
```
Comment (mblondel) [inline tests/isotonic_test.py:44]: Instead of using `mean().max()` to reduce to a scalar, you can use `self.assertArraysAllClose(output, output_sklearn)`.

File: tests/isotonic_test.py [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+    isotonic_l2_pav_vmap = jax.vmap(isotonic_l2_pav)
+    for i in range(n_batches):
+      self.assertEqual(
+          jnp.max(
+              jnp.absolute(isotonic_l2_pav_vmap(y)[i] - isotonic_l2_pav(y[i]))
+          ),
+          0,
+      )
+
+
```
Comment (mblondel) [inline tests/isotonic_test.py:67]: Remove blank line

**Review by mblondel — COMMENTED** [review]
Almost there :)

File: jaxopt/_src/isotonic.py (hunk 24 above)
Comment (michaelsdr) [reply jaxopt/_src/isotonic.py]: Done

File: tests/isotonic_test.py (hunk 25 above)
Comment (michaelsdr) [reply tests/isotonic_test.py]: Done

File: tests/isotonic_test.py (hunk 26 above)
Comment (michaelsdr) [reply tests/isotonic_test.py:61]: Done
